### PR TITLE
Add driver to allow host models (PartMC) to couple to TChem

### PR DESCRIPTION
**Pull request by jcurtis2**

### CODE REVIEW

File: src/core/TChem_Driver.cpp [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+}
+
+void TChem::Driver::freeGasKineticModel() {
+  _chem_file = std::string();
+  _therm_file = std::string();
+}
+
+void TChem::Driver::createStateVector() {
+  const ordinal_type len = TChem::Impl::getStateVectorSize(_kmcd_host.nSpec);
+  const ordinal_type nBatch = 1;
```
Comment (odiazib) [inline src/core/TChem_Driver.cpp:123]: Should we pass nBatch as an input?

File: src/core/TChem_Driver.cpp [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+  _state = real_type_2d_view_host("state dev", nBatch, len);
+  for (ordinal_type k = 0; k < len; k++) {
```
Comment (odiazib) [inline src/core/TChem_Driver.cpp:125]: This code can be replaced with Kokkos::deep_copy(_state, 0.0);.

File: src/core/TChem_Driver.cpp [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    _state(0,k) = 0.0;
+  }
+}
+
+/* Get the state vector */
+auto TChem::Driver::getStateVector() {
+  auto state_at_0 = Kokkos::subview(_state, 0, Kokkos::ALL);
```
Comment (odiazib) [inline src/core/TChem_Driver.cpp:132]: I would pass 0 as the input, e.g., ibatch.
`auto state_at_0 = Kokkos::subview(_state, ibatch, Kokkos::ALL);`

File: src/core/TChem_Driver.cpp [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+int TChem_getSpeciesName(int * index, char* result, const std::string::size_type buffer_size){
+  std::string specName = g_tchem->getSpeciesName(index);  
+  specName.copy(result, buffer_size);
+  result[specName.length()] = '\0';
+  return specName.length();
+}
+
+std::string TChem::Driver::getSpeciesName(int *index){
+  const auto speciesNamesHost = Kokkos::create_mirror_view(_kmcd_host.speciesNames);
+  Kokkos::deep_copy(speciesNamesHost, _kmcd_host.speciesNames);
```
Comment (odiazib) [inline src/core/TChem_Driver.cpp:166]: We do not need this Kokkos::deep_copy. `_kmcd_host.speciesNames` is located on the host.

File: src/core/TChem_Driver.cpp [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+/* Integrate a time step */
+void TChem::Driver::doTimestep(const double del_t){
+
+  const auto exec_space_instance = TChem::exec_space();
+//  using device_type = typename Tines::UseThisDevice<exec_space>::type;
+  using device_type = typename Tines::UseThisDevice<TChem::exec_space>::type;
+  using interf_host_device_type = typename Tines::UseThisDevice<TChem::host_exec_space>::type;
+  using problem_type = TChem::Impl::AtmosphericChemistry_Problem<real_type, interf_host_device_type>;
+  using policy_type = typename TChem::UseThisTeamPolicy<TChem::exec_space>::type;
+  policy_type policy(exec_space_instance, 1, Kokkos::AUTO());
```
Comment (odiazib) [inline src/core/TChem_Driver.cpp:203]: Here, we are assuming the number of batches is equal to 1.

Comment (odiazib) [reply src/core/TChem_Driver.cpp:203]: `Kokkos::AUTO()` does not provide the best performance. We may need to pass a team_size variable to this routine.

File: src/core/TChem_Driver.cpp [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+     }, tsum);
+     Kokkos::fence();
+  }
+  Kokkos::deep_copy(_state, state);
+}
+
+/* Return the state vector from host */
+void TChem::Driver::getStateVectorHost(real_type_2d_const_view_host &view) {
+  TCHEM_CHECK_ERROR(_state.span() == 0, "State vector should be constructed");
+  auto hv = Kokkos::create_mirror_view(_state);
```
Comment (odiazib) [inline src/core/TChem_Driver.cpp:277]: `_state` is a host view. I wonder if we need to create a mirror view.

**Review by odiazib — CHANGES_REQUESTED** [review]
It looks great! This interface assumes a batch size of 1. I recommend passing the number of batches as an input. A batch size of 1 will not provide the best performance on GPUs.

File: src/core/TChem_Driver.cpp (hunk 2 above)
Comment (jcurtis2) [reply src/core/TChem_Driver.cpp]: Resolved in 3c61a7c

File: src/core/TChem_Driver.cpp (hunk 3 above)
Comment (jcurtis2) [reply src/core/TChem_Driver.cpp]: Resolved in 3c61a7c

File: src/core/TChem_Driver.cpp (hunk 4 above)
Comment (jcurtis2) [reply src/core/TChem_Driver.cpp]: Removed deep_copy in 3c61a7c

File: src/core/TChem_Driver.cpp (hunk 6 above)
Comment (jcurtis2) [reply src/core/TChem_Driver.cpp]: Fixed in 3c61a7c

File: src/core/TChem_Driver.cpp (hunk 5 above)
Comment (jcurtis2) [reply src/core/TChem_Driver.cpp]: Fixed batch size in [42a7726](https://github.com/PCLAeroParams/TChem-atm/pull/55/commits/42a772624936477c979abcfe3345687edeb32e66). Also corrected other variables that had hardcoded batch size of 1.

File: src/core/TChem_Driver.cpp (hunk 1 above)
Comment (jcurtis2) [reply src/core/TChem_Driver.cpp]: Added as argument in [3c61a7c](https://github.com/PCLAeroParams/TChem-atm/pull/55/commits/3c61a7c28bf1d734db5fd5581c2020267aac6af7)

File: src/core/TChem_Driver.cpp (hunk 5 above)
Comment (jcurtis2) [reply src/core/TChem_Driver.cpp]: > `Kokkos::AUTO()` does not provide the best performance. We may need to pass a team_size variable to this routine.

Added something in [fe2c156](https://github.com/PCLAeroParams/TChem-atm/pull/55/commits/fe2c15617bc8091ac3b827b6903f21eb62b55783) for now where these are now able to be specified in the yaml file with other solver information.

File: src/core/TChem_Driver.cpp [hunk 7]
```diff
@@ -48,9 +49,11 @@ void initialize(const char* chemFile, const char* aeroFile, const char* numerics
 
   using ordinal_type = TChem::ordinal_type;
 
+//  ordinal_type nBatch = 1;
```
Comment (odiazib) [inline src/core/TChem_Driver.cpp]: Should this comment be deleted?

File: src/core/TChem_Driver.cpp [hunk 8]
```diff
@@ -118,38 +121,35 @@ void TChem::Driver::freeGasKineticModel() {
   _therm_file = std::string();
 }
 
-void TChem::Driver::createStateVector() {
+void TChem::Driver::createStateVector(ordinal_type nBatch) {
   const ordinal_type len = TChem::Impl::getStateVectorSize(_kmcd_host.nSpec);
-  const ordinal_type nBatch = 1;
   _state = real_type_2d_view_host("state dev", nBatch, len);
```
Comment (odiazib) [inline src/core/TChem_Driver.cpp:137]: By default, Kokkos initializes views with zeros. Thus, we do not need Kokkos::deep_copy(_state, 0.0).

File: src/core/TChem_Driver.cpp [hunk 9]
```diff
@@ -163,9 +163,9 @@ int TChem_getSpeciesName(int * index, char* result, const std::string::size_type
 
 std::string TChem::Driver::getSpeciesName(int *index){
   const auto speciesNamesHost = Kokkos::create_mirror_view(_kmcd_host.speciesNames);
-  Kokkos::deep_copy(speciesNamesHost, _kmcd_host.speciesNames);
+//  Kokkos::deep_copy(speciesNamesHost, _kmcd_host.speciesNames);
```
Comment (odiazib) [inline src/core/TChem_Driver.cpp]: Should this comment be deleted?

**Review by odiazib — APPROVED** [review]

File: src/core/TChem_Driver.cpp (hunk 9 above)
Comment (jcurtis2) [reply src/core/TChem_Driver.cpp]: This comment is outdated and addressed back in 15184967c71d9d00e7121cff2ad3647a3e2a5a48

File: src/core/TChem_Driver.cpp (hunk 7 above)
Comment (jcurtis2) [reply src/core/TChem_Driver.cpp]: This comment is outdated and was removed in 42a772624936477c979abcfe3345687edeb32e66

File: src/core/TChem_Driver.cpp (hunk 8 above)
Comment (jcurtis2) [reply src/core/TChem_Driver.cpp:137]: Good to know! I have removed the deep_copy in https://github.com/PCLAeroParams/TChem-atm/pull/55/commits/ddb3bdaeae022332cb89c615c40d9206b318600e

**Pull request merged**